rs/linux_kernel_command_line/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+/// Utilities to manipulate a kernel command line reliably.
+use std::error::Error as StdError;
+use std::fmt;
+
+#[derive(Debug)]
+/// A kernel command line with improperly-quoted argument values.
+pub struct ImproperlyQuotedValue {
+    val: String,
+}
+impl StdError for ImproperlyQuotedValue {}
+
+impl fmt::Display for ImproperlyQuotedValue {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "Unclosed quote capturing value {:?} from supplied command line",
+            self.val
+        )
+    }
+}
+
+#[derive(Debug)]
+/// A value unrepresentable as a kernel command line argument value.
+pub struct UnrepresentableValue {
+    val: String,
+}
+
+impl StdError for UnrepresentableValue {}
+
+impl fmt::Display for UnrepresentableValue {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Illegal characters in supplied value {:?}", self.val)
+    }
+}
+
+#[derive(Debug, Default)]
+/// Represents a correctly-parsed kernel command line.
+pub struct KernelCommandLine {
+    tokenized_arguments: Vec<String>,
+}
+
+impl KernelCommandLine {
+    fn format_argument(
+        argument: &str,
+        value: Option<&str>,
+    ) -> Result<String, UnrepresentableValue> {
+        Ok(argument.to_string()
```

Comment:
The code would be more readable and efficient if you split the logic based on the presence of `value` and used `format!`
For example:

```
fn escape_value(value: &str) -> Result<String, UnrepresentableValue> {
  ...
}

...
if let Some(val) = value {
  Ok(format!("{}={}", argument, Self::escape_value(val)?))
} else {
  Ok(argument.to_owned())
}
```

Reply:
Solved.